docs/react/week-2/lesson.md

```diff
@@ -364,7 +364,53 @@ function Counter() {
 }
 ```
 
-On the second render, `count` is now set to 1. Every time we click the button, the whole cycle starts again.
+On the second render, `count` is now set to 1. If we wanted to set the state to a different value, then we pass a different value to `setCount`.
+
+#### Setting state based on previous state
+
+You might have noticed that our `Counter` app isn't very useful... It can only count to 1! You might think that we could do something like this ([interactive example](https://codesandbox.io/s/usestate-counter-broken-9qd6f?file=/src/Counter.js)):
+
+:::danger
+This code has a bug!
+:::
+
+```js
+function incrementCount() {
+  setCount(count + 1);
+}
+```
+
+However, this code has a bug. Let's see what happens if we put 2 `setCount` calls within `incrementCount` ([interactive example](https://codesandbox.io/s/usestate-counter-broken-without-timeout-wel7p?file=/src/Counter.js)):
+
+```js
+function incrementCount() {
+  setCount(count + 1);
+  setCount(count + 1);
+}
+```
+
+If you click the button, then the counter will only count up to 1. This is because the `count` variable does not actually update until the component re-renders and `useState` gives a new value of `count`. We can see this if we put a `console.log(count)` in-between the `setCount` calls.
+
+As you can see, `count` remains at 0, even after we try to update it to 1. This means that in the second `setCount` call we are actually still trying to update the state to 1!
+
+To fix this problem, there is another way we can call `setCount` if we always need to get the latest version of state when setting it ([interactive example](https://codesandbox.io/s/usestate-counter-updating-based-on-previous-state-dyyo6?file=/src/Counter.js)):
+
+```js
+function incrementCount() {
+  setCount((c1) => {
+    return c1 + 1;
+  });
+  setCount((c2) => {
+    return c2 + 1;
+  });
+}
+```
+
+When we pass a callback function to `setCount`, React will pass us the **latest version** of the count state. In this example, the variables are named `c1` and `c2` to make it clear that they are different from the `count` variable. Now if we add `console.log(c1)` and `console.log(c2)` then we can see that they get the correct values.
```

Comment:
```suggestion
When we pass a callback function to `setCount`, React will pass us the **latest version** of the count state. In this example, the variables are named `c1` and `c2` to make it clear that they are different from the `count` variable. Now if we add `console.log(c1)` and `console.log(c2)` then we can see that they get the correct values.

Reminder: `c1` and `c2` are just arbitrary names we pick to be the name of the parameter to the function we're passing - we could have called them both `c`, and everything would still work - they're not related to each other, as they are defined in different scopes.
```

As they tend to struggle with scope, I'm a little concerned that without this clarification, trainees may think that these need to be different, or have some relation to each other.

Reply:
Yeah, good suggestion thanks.